File: server/tools/thinking.py
"""Split / strip model <think> and <thinking> blocks from assistant text."""
from __future__ import annotations

import re

# Closed pairs first; tag name must match on close.
_THINK_PAIR = re.compile(
    r"<(think|thinking)>([\s\S]*?)</\1>",
    re.IGNORECASE,
)
_THINK_OPEN = re.compile(r"<(think|thinking)>", re.IGNORECASE)
_THINK_CLOSE = re.compile(r"</(?:think|thinking)>", re.IGNORECASE)


def split_thinking(text: str) -> tuple[str, str]:
    """Split assistant text into (content, thinking).

    - Recognizes ``<think>`` and ``<thinking>`` (case-insensitive).
    - Merges all think blocks into one string (joined by blank lines).
    - Unclosed open tag: everything after it is treated as thinking (not content).
    """
    if not text:
        return "", ""

    parts: list[str] = []

    def _collect(match: re.Match[str]) -> str:
        chunk = (match.group(2) or "").strip()
        if chunk:
            parts.append(chunk)
        return ""

    content = _THINK_PAIR.sub(_collect, text)
    open_m = _THINK_OPEN.search(content)
    if open_m:
        rest = content[open_m.end() :].strip()
        if rest:
            parts.append(rest)
        content = content[: open_m.start()]
    content = _THINK_CLOSE.sub("", content).strip()
    thinking = "\n\n".join(parts)
    return content, thinking


def strip_thinking(text: str) -> str:
    """Return content only (thinking discarded)."""
    return split_thinking(text)[0]

File: ui-apple/src/utils/thinking.js
/**
 * Split assistant text into { content, thinking }.
 * Recognizes <think> and <thinking> (case-insensitive).
 * Merges all blocks; unclosed open tag → rest is thinking.
 */

const THINK_PAIR = /<(think|thinking)>([\s\S]*?)<\/\1>/gi
const THINK_OPEN = /<(think|thinking)>/i
const THINK_CLOSE = /<\/(?:think|thinking)>/gi

export function splitThinking(text) {
  if (!text) return { content: '', thinking: '' }

  const parts = []
  let content = String(text).replace(THINK_PAIR, (_m, _tag, inner) => {
    const chunk = (inner || '').trim()
    if (chunk) parts.push(chunk)
    return ''
  })

  const openM = content.match(THINK_OPEN)
  if (openM) {
    const idx = content.search(THINK_OPEN)
    const rest = content.slice(idx + openM[0].length).trim()
    if (rest) parts.push(rest)
    content = content.slice(0, idx)
  }

  content = content.replace(THINK_CLOSE, '').trim()
  const thinking = parts.join('\n\n')
  return { content, thinking }
}

export function stripThinking(text) {
  return splitThinking(text).content
}
